### DISCUSSION TITLE
Is steel scheme good as a system scripting language?

### DISCUSSION THREAD

**Discussion by amano-kenji** [mattwparas/steel#148] [Q&A] [answered]
Janet has been the best system scripting langauge for me.

By default, it uses `/usr/lib(64)/janet` as the directory for modules, but users can set janet's module path....

steel scheme doesn't set STEEL_HOME by default.

For a language to be used as a system scripting language, the defualt module directory must be provided.

**Comment by mattwparas** [comment]
I would argue that it does not _have_ to be provided by default, assuming installation instructions are sufficient. The only reason I have not assumed a default yet is because I don't distribute (up to date) binaries - most if not all are building from source. Of course, at some point soon I will pick a default directory that can be overridden. The engine itself has support for adding other directories to search for modules, it just isn't directly exposed in the repl.

**Reply by amano-kenji** [reply]
Linux distributions will have packaging scripts that build and install helix and steel.

**Reply by mattwparas** [reply]
Sure, but steel isn't yet merged into helix, so I don't think its a huge issue at the current moment. If you feel strongly about this, feel free to open an issue

**Reply by amano-kenji** [reply]
Is it safe to assume that steel's priority is being embedded into other programs? Janet can be also be easily embedded, but it has been used as a standalone scripting system.

**Reply by mattwparas** [reply]
I'm not sure why we need to directly compare Janet and Steel - yes, they have similar properties, but Janet is older and perhaps has different goals. I don't follow it too much, however I have a good deal of admiration for the project.

The implementation of steel is designed to be flexible, either embedded or standalone. I do not particularly prioritize one over the other; I use it both ways.

**Reply by amano-kenji** [reply]
It seems steel is in almost the same niche as janet...

**Reply by mattwparas** [reply]
yes, they may do similar things, but nowhere have I said "use steel instead of janet" - if you don't want to use steel, you don't have to. This isn't a competition - I work on steel because I enjoy doing it, whether or not there are users. If janet suits your needs more, please continue to use it by all means, it is a very cool project.

**Reply by amano-kenji** [reply]
In janet, lack of persistent data structure and precision math support bothers me, but I have been able to make do with janet in most cases. Precision math library can be added to janet since it already has a big integer library.

I don't know whether persistent data structures can be added to janet core without making it bloated.

**Reply by mattwparas** [reply]
Well, how do you use Janet? Janet is written in C designed for embedding in C applications (if I'm remembering correctly) - Steel is mostly designed for Rust application in mind. I don't think they're directly comparable in that way

**Reply by amano-kenji** [reply]
I think janet is not embedded in anything, yet. It is used as a standalone system scripting language. It is primarily a system scripting language. It is marketed as a system scripting language that's also easy to embed.

I use it as a system scripting language. I don't really understand C. I don't know Rust.

If I'm going to use steel scheme in helix anyway, I may consider using steel scheme as a standalone system scripting language as well.

If steel scheme starts up fast and is as mature as janet, I may even consider moving to steel scheme because steel scheme has persistent data structures which janet doesn't have. Janet has a third party persistent data structure library which still doesn't prevent threads from copying when I pass values between threads. Built-in persistent data structures can prevent threads from copying values when values are passed between threads... Persistent data structures are also easier to work with than immutable data structures... in my experiences...

**Reply by amano-kenji** [reply]
I use janet mostly for interactive scripts linked to keyboard shortcuts. Because janet starts up and quits in 5 milliseconds, it is good for keyboard shortcuts. I also use it as a better alternative to bash and python.

janet's parsing expression grammar(PEG) is far better than regular expression and awk. steel scheme should have something like PEG.

It is also ergonomic for general scripting tasks. I wrote i3bar/swaybar scripting system in janet.... It is remarkably ergonomic for scripting i3bar/swaybar. Channels are used to pass values between asynchronous tasks. Threaded channels are used to copy and pass values between threads. Channels are ergonomic for concurrency and parallelism. Asynchronous programming in raku is painful on the other hand...

I expect steel scheme to be similar in that regard if it matures...

**Reply by mattwparas** [reply]
To address a few of those points:

* Steel takes about 20-30 ms to startup, however with some more effort I can get that down to < 10 ms. Just requires a little work I haven't had the time to allocate in the last few weeks
* Steel isn't as mature as Janet yet, like I said Janet has a few years on Steel
* The threading model in Steel does deep copy values when passing between threads. I will eventually have a notion of persistent data structures that are safe to share between threads, but as of now most values are not thread safe, which means we have to do a copy.
* "steel scheme should have something like PEG" - if you're willing to contribute something, I'm happy to accept contributions, but this is not high on my priority list.

I am not attempting to compete or siphon users from Janet. If Janet has a feature set that you appreciate, you should use that. If Steel does not have that feature set, and you would like it, please open an issue and describe what you'd like. If I think its a reasonable feature, I will consider adding it, and am always happy to accept contributions as well.

**Reply by amano-kenji** [reply]
Wait, steel scheme doesn't have regular expression, parsing expression grammar, or something else for text parsing?

**Reply by amano-kenji** [reply]
Which types does steel scheme use or plan to use as numbers?

**Reply by mattwparas** [reply, marked as answer]
I have not yet implemented the regex library, however the plan is to wrap the rust regex library https://docs.rs/regex/latest/regex/ and include it with the interpreter.

Steel will be implementing the full numeric tower, which currently includes integers (isize), floats (f64), big ints, and will eventually include rationals as well

**Reply by amano-kenji** [reply]
My recommendation is to skip regex and go straight to parsing expression grammar.

Raku's grammar is an evolution of regex, but parsing expression grammar is superior to even raku's grammar in my experiences.

After having used regex, raku grammar, and PEG, I think PEG is the best.

**Reply by amano-kenji** [reply]
I found https://github.com/kevinmehall/rust-peg, but I don't know how to write in Rust.

**Reply by amano-kenji** [reply]
Is the rational number based on big int? If it is, then it can be an arbitrary-precision decimal number.

In Raku, Rat type has arbitrary-size integer numerator and 64-bit denominator. Raku's FatRat type gives you arbitrary-size integer for denominator.

https://hackage.haskell.org/package/Decimal is expressed by arbitrary-size integer and 8-bit exponent. This means you can have precision up to 255 decimal places with Haskell's Decimal. Enough for financial calculation.

**Reply by novusnota** [reply]
+1 for PEG support, as an alternative or in addition to the regex

**Reply by saolof** [reply]
I would personally be a bit hesitant about PEG unless the implementation is carefully designed simply because of how easily it can turn into a mess. Compared to recursive descent you reduce boilerplate slightly but lose a lot of explicitness. A few basic helper functions to remove boilerplate from tokenizers (for example, convenience functions to split by whitespace and unicode character category) and recursive descent (control flow conveniences, pull/push functions in transducers, etc) may be a better approach.

The main case where a parser DSL is useful is if it composes well with something else, like for example if the parser allows pushing tokens into it instead of just pulling (reasonably straightforward for table-based parsers or with coroutines/callcc), since that makes it pipeline-friendly. Or having a good parser combinator DSL that can be tested piecewise.

**Reply by amano-kenji** [reply]
Parser is a fine alternative to PEG, but PEG is a better alternative to regular expression which is a mess.
